cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -82,52 +88,46 @@ type etcdClient interface {
 	// MemberPromote promotes a member from raft learner (non-voting) to raft voting member.
 	MemberPromote(ctx context.Context, id uint64) (*clientv3.MemberPromoteResponse, error)
 
-	// Endpoints lists the registered endpoints for the client.
-	Endpoints() []string
-
-	// Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
-	Sync(ctx context.Context) error
-
 	// Status gets the status of the endpoint.
 	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
 
-	// Close shuts down the client's etcd connections.
-	Close() error
+	// Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
+	Sync(ctx context.Context) error
 }
 
 // Client provides connection parameters for an etcd cluster
 type Client struct {
 	Endpoints []string
-	TLS       *tls.Config
 
-	newEtcdClient func() (etcdClient, error)
+	newEtcdClient func(endpoints []string) (etcdClient, error)
 }
 
 // New creates a new EtcdCluster client
 func New(endpoints []string, ca, cert, key string) (*Client, error) {
 	client := Client{Endpoints: endpoints}
 
+	var err error
+	var tlsConfig *tls.Config
 	if ca != "" || cert != "" || key != "" {
 		tlsInfo := transport.TLSInfo{
 			CertFile:      cert,
 			KeyFile:       key,
 			TrustedCAFile: ca,
 		}
-		tlsConfig, err := tlsInfo.ClientConfig()
+		tlsConfig, err = tlsInfo.ClientConfig()
 		if err != nil {
 			return nil, err
 		}
-		client.TLS = tlsConfig
 	}
 
-	client.newEtcdClient = func() (etcdClient, error) {
+	client.newEtcdClient = func(endpoints []string) (etcdClient, error) {
```

Comment:
what's the benefit of passing the `endpoints` here vs. the original way to set the field of  `Endpoints` by `client.Endpoints`?

Reply:
Good question.

I [instantiate](https://github.com/kubernetes/kubernetes/blob/44819c18482ff1fae243df4210488b824e838d38/cmd/kubeadm/app/util/etcd/etcd.go#L123) `client.newEtcdClient` using an anonymous closure.

Originally, I _did_ reference the struct member `Endpoints` from the closure. However, the `Endpoints` member [is updated by the `Sync` receiver](https://github.com/kubernetes/kubernetes/blob/44819c18482ff1fae243df4210488b824e838d38/cmd/kubeadm/app/util/etcd/etcd.go#L262). I felt that it was not clear to the reader that the closure would read the up-to-date value of `Endpoints`.

I thought that explicitly passing the value to the closure would make it  clear that the closure uses the up-to-date value of `Endpoints`.

Reply:
Ack.